cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -39,7 +39,7 @@ Status ValidateJoinInputs(const std::shared_ptr<Schema>& left_schema,
                           const std::shared_ptr<Schema>& right_schema,
                           const std::vector<int>& left_keys,
                           const std::vector<int>& right_keys) {
-  if (left_keys.size() != right_keys.size()) {
+  if (left_keys.size() != right_keys.size() && left_keys.size() > 0) {
```

Comment:
Is it valid to join with no keys?

Reply:
AFAIK I don't think it's valid. We'd need some indexer if no key columns are specified

Reply:
Sorry I replied by email and it seems to have gotten messed up. Then maybe this should be 

```
if (left_keys.size() == 0) { return Status::Invalid(...); }
if (left_keys.size() != right_keys.size())) { ...}
```

?

Reply:
Sorry I missed this. I added the changes now